docker-entrypoint.sh

```diff
@@ -1,117 +1,584 @@
+      edbdocker_log "Bootstrap script $filename"
+      cat "$filename" | edgedb "${conn_opts[@]}"
+    done
+  fi
+}
+
+
+_edbdocker_bootstrap_abort_cb() {
+  if [ -n "${EDGEDB_DATADIR}" ] && [ -e "${EDGEDB_DATADIR}" ]; then
+    (shopt -u nullglob; rm -rf "${EDGEDB_DATADIR}/"* || :)
+  fi
+}
+
+
+# Runs schema migrations found in /dbschema unless EDGEDB_SKIP_MIGRATIONS
+# is set.  Expects either EDGEDB_DATADIR or EDGEDB_POSTGRES_DSN to be set
+# in the environment.  Migrations are applied by a temporary edgedb-server
+# process that gets started on a random port and is shut down once bootstrap
+# is complete.
+#
+# Usage: `EDGEDB_DATADIR=/foo/bar edbdocker_run_migrations`
+edbdocker_run_migrations() {
+  if [ -d "/dbschema" ] && [ -z "${EDGEDB_SKIP_MIGRATIONS:-}" ]; then
+    local server_opts
+
+    server_opts=()
+    if [ -n "${EDGEDB_POSTGRES_DSN}" ]; then
+      server_opts+=(--postgres-dsn="${EDGEDB_POSTGRES_DSN}")
+    else
+      server_opts+=(--data-dir="${EDGEDB_DATADIR}")
     fi
-fi
 
+    edbdocker_log "Applying schema migrations..."
+    edbdocker_run_temp_server \
+      _edbdocker_migrations_cb \
+      _edbdocker_migrations_abort_cb \
+      "${server_opts[@]}"
+  fi
+}
+
+
+_edbdocker_migrations_cb() {
+  if ! edgedb "${@}" migrate --schema-dir=/dbschema; then
+    edbdocker_log "ERROR: Migrations failed. Stopping server."
+    return 1
+  fi
+}
+
+
+_edbdocker_migrations_abort_cb() {
+  :
+}
+
+
+# Write arguments to stderr separated by newlines.
+# Example: `edbdocker_log "some" "long" "message"`
+# will output:
+#   some
+#   long
+#   message
+#
+edbdocker_log() {
+  printf >&2 "%s\n" "${@}"
+}
+
+
+# Log arguments to stderr using `edbdocker_log` and exit with 1.
+edbdocker_die() {
+  edbdocker_log "${@}"
+  exit 1
+}
+
+
+# Start edgedb-server on a random port, execute the specified callback
+# and shut down the server.
+#
+# Usage: `edbdocker_run_temp_server callback abort_callback --server-arg=val ...`
+edbdocker_run_temp_server() {
+  local edgedb_pid
+  local shell_pid
+  local timeout_pid
+  local timeout
+  local retry_period
+  local runstate_dir
+  local port
+  local ecode
+  local server_opts
+  local conn_opts
+  local callback
+
+  callback="$1"
+  abort_callback="$2"
+  shift 2
+  server_opts=( "${@}" )
+
+  runstate_dir="$(mktemp -d)"
+  server_opts+=(
+    --runstate-dir="$runstate_dir"
+    --port="auto"
+    --bind-address="127.0.0.1"
+  )
+  conn_opts=( --admin --host="$runstate_dir" --wait-until-available="2s" )
 
-exec "$@"
+  # Start the server
+  edgedb-server "${server_opts[@]}" &
+  edgedb_pid="$!"
+
+  timeout=120
+  retry_period=5
+
+  function _abort() {
+    $abort_callback
+    if [ -e "${runstate_dir}" ]; then
+      rm -r "${runstate_dir}" || :
+    fi
+    edbdocker_die "$1"
+  }
+
+  (
+    local port
+    local sock
+    local i
+    local check
+
+    shopt -s nullglob  # to properly match the Unix socket
+
+    for i in $(seq 1 $(($timeout / $retry_period))); do
```

Comment:
Why not CLI's `--wait-until-available`. It's specifically implemented for this use case? And will react faster than in 5 seconds.

Reply:
We can't use `--wait-until-available` here, because we don't know the port.  I considered using `--echo-runtime-info`, but it didn't seem less messy.

Reply:
IMO, the nicest approach would be to teach `edgedb server` to start a server given its full command line and wait for it to start properly.  Same for stopping/etc.  Basically our version of `pg_ctl`.  Useful here and for where there isn't systemd to manage the service for us.